Source/DafnyCore/Resolver/PreType/PreType.cs

```diff
@@ -0,0 +1,326 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Bpl = Microsoft.Boogie;
+
```

Comment:
Some big picture comments explaining what pre-types are about and what the classes and methods do would be welcome

Reply:
I added a comment in the code. In essence, pre-types are the part of types that is mutually dependent on name resolution. This means a pre-type expands type synonyms and subset types. (See also the comments at the top of this PR.)